vllm/model_executor/models/phi_1_5.py

```diff
@@ -0,0 +1,274 @@
+        return hidden_states
+
+
+class PhiModel(nn.Module):
+
+    def __init__(self, config: PretrainedConfig):
+        super().__init__()
+        self.config = config
+
+        self.wte = VocabParallelEmbedding(
+            config.vocab_size,
+            config.n_embd,
+        )
+        self.layers = nn.ModuleList(
+            [PhiLayer(config) for _ in range(config.n_layer)])
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        position_ids: torch.Tensor,
+        kv_caches: List[KVCache],
+        input_metadata: InputMetadata,
+        cache_events: Optional[List[torch.cuda.Event]],
+    ) -> torch.Tensor:
+        hidden_states = self.wte(input_ids)
+        for i in range(len(self.layers)):
+            if cache_events is None:
+                cache_event = None
+            else:
+                cache_event = cache_events[i]
+            layer = self.layers[i]
+            hidden_states = layer(
+                position_ids,
+                hidden_states,
+                kv_caches[i],
+                input_metadata,
+                cache_event,
+            )
+        return hidden_states
+
+
+class PhiForCausalLM(nn.Module):
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.phi = PhiModel(config)
+        self.ln = nn.LayerNorm(config.n_embd, eps=config.layer_norm_epsilon)
+        self.linear = ColumnParallelLinear(
+            config.n_embd,
+            config.vocab_size,
+            gather_output=False,
+        )
+        self.sampler = Sampler(config.vocab_size)
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[KVCache],
+        input_metadata: InputMetadata,
+        cache_events: Optional[List[torch.cuda.Event]],
+    ) -> SamplerOutput:
+        hidden_states = self.phi(input_ids, positions, kv_caches,
+                                 input_metadata, cache_events)
+        hidden_states = self.ln(hidden_states)
+        next_tokens = self.sampler(self.linear.weight, hidden_states,
+                                   input_metadata, self.linear.bias)
+        return next_tokens
+
+    _column_parallel_weights = [
+        "embed_in.weight", "embed_out.weight", "embed_out.bias", "fc1.weight",
+        "fc1.bias"
+    ]
+    _row_parallel_weights = ["out_proj.weight", "fc2.weight"]
+
+    def load_weights(self,
+                     model_name_or_path: str,
+                     cache_dir: Optional[str] = None,
+                     load_format: str = "auto",
+                     revision: Optional[str] = None):
+        tensor_model_parallel_rank = get_tensor_model_parallel_rank()
+        state_dict = self.state_dict()
+        for name, loaded_weight in hf_model_weights_iterator(
+                model_name_or_path, cache_dir, load_format, revision):
+            if "rotary_emb.inv_freq" in name:
```

Comment:
BTW, why is the `inv_freq` here different from the one in our RotaryEmbedding? According to this [line](https://huggingface.co/microsoft/phi-1_5/blob/92557d03bb12543040c8bb5f0475cbdd9968f05f/modeling_mixformer_sequential.py#L258), the code looks the same.

Reply:
Yeah, I also thought so until I checked the weight:

- `inv_freq` from `RotaryEmbedding`
```
tensor([1.0000e+00, 5.6234e-01, 3.1623e-01, 1.7783e-01, 1.0000e-01, 5.6234e-02,
        3.1623e-02, 1.7783e-02, 1.0000e-02, 5.6234e-03, 3.1623e-03, 1.7783e-03,
        1.0000e-03, 5.6234e-04, 3.1623e-04, 1.7783e-04], device='cuda:0',
       dtype=torch.float32)
```
- `loaded_weight` of `rotary_emb.inv_freq`
```
tensor([1.0000e+00, 5.6250e-01, 3.1616e-01, 1.7786e-01, 9.9976e-02, 5.6244e-02,
             3.1616e-02, 1.7776e-02, 1.0002e-02, 5.6229e-03, 3.1624e-03, 1.7786e-03,
             1.0004e-03, 5.6219e-04, 3.1614e-04, 1.7786e-04])
```

But even more surprising is that after I copied the cached weight, the output on `Test 4` was still different. However, @Linzecong says that it might be an issue of fp16 [link](https://github.com/vllm-project/vllm/issues/1167#issuecomment-1811733712)

Reply:
After some investigations, I found that the issue is because `inv_freq` is stored in FP16 in the weight checkpoint (I guess it is calculated in FP32 and converted to FP16 for some reason). As we use the same logic to calculate `inv_freq`, I think this slight difference should be acceptable.

Can we remove the special weight loading logic for `inv_freq` and use our current implementation instead?